### PR TITLE
Implement Auth service`s interface

### PR DESCRIPTION
**Pull request by maza4aka**
* Auth implemented as singleton, same way as Storage
* AuthImpl has only some stubs for red tests (TODO: implement methods` logic)
* Some user login/creation (red) tests

### CODE REVIEW

File: src/main/java/edu/kpi/testcourse/auth/AuthFactory.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+public class AuthFactory {
+  private static Auth auth = null;
+
+  /**
+   * Creates and/or returns Authentication service singleton.
+   *
+   * @return Auth instance
+   */
+  @Singleton
+  Auth auth() {
```
Comment (AlexPraefectus) [inline src/main/java/edu/kpi/testcourse/auth/AuthFactory.java:19]: This implementation is not thread-safe and can cause lots of problems in multithreaded environments.

Please take a look at our discussion with @OlehKyba https://github.com/future-stardust/url-shrtnr-mystandstarplatinum/pull/3#pullrequestreview-606196942, there was a similar problem.

TL;DR you can make the method `synchronized to avoid problems.

File: src/main/java/edu/kpi/testcourse/auth/Auth.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  AuthStatus createUser(String username, String password);
+
+  /**
+   * Validates user login.
+   *
+   * @param username user`s email
+   * @param password user`s password
+   * @return auth service status
+   */
+  AuthStatus loginUser(String username, String password);
```
Comment (AlexPraefectus) [inline src/main/java/edu/kpi/testcourse/auth/Auth.java:24]: I think that `loginUser` method should return JWT on success 🤔 

Unfortunately, we're not in Rust and are unable to bind types to enum values.  

Maybe, the most Java way is to return token and to throw custom Exceptions on errors.

File: src/main/java/edu/kpi/testcourse/auth/Auth.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+public interface Auth {
+
+  /**
+   * Creates new user.
+   *
+   * @param username user`s email
+   * @param password user`s password
+   * @return auth service status
+   */
+  AuthStatus createUser(String username, String password);
```
Comment (AlexPraefectus) [inline src/main/java/edu/kpi/testcourse/auth/Auth.java:15]: Returning enum value is legit but maybe this method should return void and throw Exceptions on errors?

I think that it will be more Java way of writing code and the interface of this method will also be similar to `loginUser` (after it will be changed)

**Review by AlexPraefectus — CHANGES_REQUESTED** [review]
Please take a look at the proposed changes.  
Let's discuss if something is not clear or You have any questions regarding my comments.

File: src/main/java/edu/kpi/testcourse/auth/AuthFactory.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package edu.kpi.testcourse.auth;
+
+import io.micronaut.context.annotation.Factory;
+import javax.inject.Singleton;
+
+/**
+ * Factory to create Authentication service.
+ */
+@Factory
+public class AuthFactory {
```
[the remark this answers is not in the document]
Comment (maza4aka) [reply src/main/java/edu/kpi/testcourse/auth/AuthFactory.java:10]: Idk, your Storage factory also produces singleton instance, what's the difference? Anyway, I`m happy that it works, at least.

File: src/main/java/edu/kpi/testcourse/auth/AuthFactory.java (hunk 1 above)
Comment (maza4aka) [reply src/main/java/edu/kpi/testcourse/auth/AuthFactory.java:19]: Auth instance is created when initAuth is called inside Main method, not sure if it will cause problems. As for reading this object from multiple threads, it is still safe (except for directly working with Auth instance from multiple threads, which I doubt will eventually happen anyway).

File: src/main/java/edu/kpi/testcourse/auth/Auth.java (hunk 2 above)
Comment (maza4aka) [reply src/main/java/edu/kpi/testcourse/auth/Auth.java:24]: Framework will do this automatically, as I understand. All I have to do is return whether user can login or not.

File: src/main/java/edu/kpi/testcourse/auth/Auth.java (hunk 3 above)
Comment (maza4aka) [reply src/main/java/edu/kpi/testcourse/auth/Auth.java:15]: Using enums to tell what's happened is much easier and cleaner for me. Also, interface for these two mentioned methods is already pretty consistent.

If you want to use exceptions, good for you, that's a matter of taste - propose your changes via PR, if you want.

**Comment by maza4aka** [comment]
Auth service is almost implemented, only password hashing and Auth usage by AuthenticationProviderUserPassword is left for another PR.

File: src/main/java/edu/kpi/testcourse/auth/AuthFactory.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (AlexPraefectus) [reply src/main/java/edu/kpi/testcourse/auth/AuthFactory.java:10]: I used factory pattern to avoid many threads trying to initialize storage (which requires reading a lot of data from disk). And it also worked like barrier sync to wait until storage will be initialized.  
In this case, You can simply mark `Auth` class with `@Singleton` annotation and it will work the same way as it works now without factory.

[the remark this answers is not in the document]
Comment (AlexPraefectus) [reply src/main/java/edu/kpi/testcourse/auth/AuthFactory.java:10]: But as I have already stated before, current implementation with factory injection is pretty much good too :slightly_smiling_face:

File: src/main/java/edu/kpi/testcourse/auth/AuthFactory.java (hunk 1 above)
Comment (AlexPraefectus) [reply src/main/java/edu/kpi/testcourse/auth/AuthFactory.java:19]: It's very implicit contract IMHO.  
And marking method `synchronized`  will be a very "cheap" fix.

File: src/main/java/edu/kpi/testcourse/auth/AuthImpl.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+   * @param username username
+   * @return InvalidUsername or Ok
+   */
+  private AuthStatus checkUsername(String username) {
+    if (username == null) {
+      return AuthStatus.InvalidUsername;
+    }
+    if (username.isEmpty()) {
+      return AuthStatus.InvalidUsername;
+    }
```
Comment (AlexPraefectus) [inline src/main/java/edu/kpi/testcourse/auth/AuthImpl.java:49]: this block can be simplified using one `if`:
```java
if (username == null || username.isEmpty()) {
  return AuthStatus.InvalidUsername;
}
```

File: src/main/java/edu/kpi/testcourse/auth/AuthImpl.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+   * @param password password
+   * @return InvalidPassword or Ok
+   */
+  private AuthStatus checkPassword(String password) {
+    if (password == null) {
+      return AuthStatus.InvalidPassword;
+    }
+    if (password.isEmpty()) {
+      return AuthStatus.InvalidPassword;
+    }
```
Comment (AlexPraefectus) [inline src/main/java/edu/kpi/testcourse/auth/AuthImpl.java:65]: Same as username - can be simplified.  
Also, there's a DRY violation - String checks are absolutely the same for `username` and `password`

Maybe it would be a good idea to introduce method kinda `private bool checkStringNotEmpty(String s)` and move these checks there.

File: src/main/java/edu/kpi/testcourse/auth/AuthImpl.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+  /**
+   * Check that user exists.
+   *
+   * @param username username
+   * @return true if user exists
+   */
+  public boolean isUserExists(String username) {
+    try {
+      return authStorage.get(username) != null;
+    } catch (NullPointerException e) {
```
Comment (AlexPraefectus) [inline src/main/java/edu/kpi/testcourse/auth/AuthImpl.java:101]: as far as I remember, storage never throws

File: src/main/java/edu/kpi/testcourse/auth/AuthImpl.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+    if ((usernameStatus != AuthStatus.Ok) && (passwordStatus != AuthStatus.Ok)) {
+      return AuthStatus.InvalidCredentials;
+    }
+    if (usernameStatus != AuthStatus.Ok) {
+      return AuthStatus.InvalidUsername;
+    }
+    if (passwordStatus != AuthStatus.Ok) {
+      return AuthStatus.InvalidPassword;
+    }
```
Comment (AlexPraefectus) [inline src/main/java/edu/kpi/testcourse/auth/AuthImpl.java:88]: Maybe it will be a good idea to simply propagate `usernameStatus` and `passwordStatus`

File: src/main/java/edu/kpi/testcourse/auth/AuthImpl.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+  /**
+   * Check user`s credentials correctness.
+   *
+   * @param username username
+   * @param password password
+   * @return AuthStatus value
+   */
+  private AuthStatus validateUser(String username, String password) {
+    if (!isUserExists(username)) {
+      return AuthStatus.InvalidUsername;
```
Comment (AlexPraefectus) [inline src/main/java/edu/kpi/testcourse/auth/AuthImpl.java:131]: User that does not exist and invalid data in the username should be different errors i think. With current implementation we're unable to process these errors separately (return different HTTP error codes for example)

File: src/main/java/edu/kpi/testcourse/auth/AuthImpl.java (hunk 5 above)
Comment (maza4aka) [reply src/main/java/edu/kpi/testcourse/auth/AuthImpl.java:49]: ok, indeed

File: src/main/java/edu/kpi/testcourse/auth/AuthImpl.java (hunk 6 above)
Comment (maza4aka) [reply src/main/java/edu/kpi/testcourse/auth/AuthImpl.java:65]: alright

File: src/main/java/edu/kpi/testcourse/auth/AuthImpl.java (hunk 7 above)
Comment (maza4aka) [reply src/main/java/edu/kpi/testcourse/auth/AuthImpl.java:101]: it throws NullPointer if username is null

File: src/main/java/edu/kpi/testcourse/auth/AuthImpl.java (hunk 8 above)
Comment (maza4aka) [reply src/main/java/edu/kpi/testcourse/auth/AuthImpl.java:88]: yup, ok

File: src/main/java/edu/kpi/testcourse/auth/AuthImpl.java (hunk 9 above)
Comment (maza4aka) [reply src/main/java/edu/kpi/testcourse/auth/AuthImpl.java:131]: I think, for security reasons, nobody outside should know if the user exists - if they provide some username and get InvalidUsername, no one will know if user even exists

File: src/main/java/edu/kpi/testcourse/auth/AuthImpl.java (hunk 7 above)
Comment (AlexPraefectus) [reply src/main/java/edu/kpi/testcourse/auth/AuthImpl.java:99]: Well, this is invalid data for storage and can cause things to break if internal container will be changed from `Hashtable` to `HashMap` or something like this.
I think, it's better to check that data is not null before trying to use it with storage.

File: src/main/java/edu/kpi/testcourse/auth/AuthFactory.java (hunk 1 above)
Comment (maza4aka) [reply src/main/java/edu/kpi/testcourse/auth/AuthFactory.java:19]: ok, why not

File: src/main/java/edu/kpi/testcourse/auth/AuthImpl.java (hunk 7 above)
Comment (maza4aka) [reply src/main/java/edu/kpi/testcourse/auth/AuthImpl.java:99]: but that's the contract your interface provides, at least in documentation. I shouldn't rely on it?

Comment (AlexPraefectus) [reply src/main/java/edu/kpi/testcourse/auth/AuthImpl.java:99]: Oh, sorry, my bad, forgot that this is documented behaviour

File: src/main/java/edu/kpi/testcourse/auth/AuthImpl.java (hunk 9 above)
Comment (AlexPraefectus) [reply src/main/java/edu/kpi/testcourse/auth/AuthImpl.java:129]: Sounds reasonable :+1:

**Review by AlexPraefectus — APPROVED** [review]
LGTM :+1:

**Comment by maza4aka** [comment]
@AlexPraefectus, I will restore this branch, if you don't mind) I will use it for the next two PRs.

**Comment by AlexPraefectus** [comment]
Feel free to, it will be auto restored on your next push)
Just a habit to delete branches after merge)
